### PR TITLE
Allow converting sliced bytes to uint256

### LINKED ISSUE #1072: Converting sliced bytes to non-bytes32 doesn't work
### Version Information
* vyper Version: 0.1.0b4

### What's your issue about?
Vyper doesn't like converting a slice of `bytes` (slice len <=32) to a type other than `bytes32`:

```python
# Compiles!
@public
@constant
def foo(_arg: bytes[100]) -> bytes32:
    return convert(slice(_arg, start=20, len=32), bytes32)

# Does not compile....
@public
@constant
def bar(_arg: bytes[100]) -> uint256:
    return convert(slice(_arg, start=20, len=32), uint256)

# Compiles! (Workaround)
@public
@constant
def baz(_arg: bytes[100]) -> uint256:
    return convert(convert(slice(_arg, start=20, len=32), bytes32), uint256)

# Note: above does not work for addresses!
```

### How can it be fixed?
Make it so!

#### Cute Animal Picture

fluffy pony

### PR DESCRIPTION
**Pull request by jakerockland**
### - What I did

Fixes #1072 (this may currently be only a partial fix, pending discussion on issue)

### - How I did it

Updated `convert.py` to allow converting to `uint256` using existing internal `byte_array_to_num` method that supports sliced bytes. Added relevant test to `test_uint256.py`.

### - How to verify it

`make test`

### - Description for the changelog

Added support for converting sliced `bytes` to `uint256`.

### - Cute Animal Picture

image

### CODE REVIEW

File: vyper/types/convert.py [hunk 1]
```diff
@@ -60,6 +60,9 @@ def to_uint256(expr, args, kwargs, context):
     elif isinstance(in_node, LLLnode) and input_type in ('bytes32', 'address'):
         return LLLnode(value=in_node.value, args=in_node.args, typ=BaseType('uint256'), pos=getpos(expr))
 
+    elif isinstance(in_node, LLLnode) and input_type is 'bytes':
+        return byte_array_to_num(in_node, expr, 'uint256')
```
[the remark this answers is not in the document]
Comment (jakerockland) [reply vyper/types/convert.py]: Rad, that SGTM.

**Comment by jakerockland** [comment]
@fubuloubu Just pushed a fix per the review you requested. In process I also noticed that there is a bug right now with conversion from `bytes` to `int128` which allows something like this to compile:

```python
# Convert bytes array of > 32 bytes to int128 _should_ fail
@public
def foo(bar: bytes[40]) -> int128:
    return convert(bar, int128)
```

Opening a separate PR that I believe fixes this.

File: tests/parser/types/numbers/test_uint256.py [hunk 2]
```diff
@@ -1,17 +1,44 @@
...
 
-    c = get_contract_with_gas_estimation(test_contract)
+    c = get_contract_with_gas_estimation(test_success)
     assert c.foo(b'\x00\x00\x00\x00\x00') == 0
     assert c.foo(b'\x00\x07\x5B\xCD\x15') == 123456789
 
+    # Test overflow bytes input for conversion
+    test_fail = """
+@public
+def foo(bar: bytes[40]) -> uint256:
```
Comment (fubuloubu) [inline tests/parser/types/numbers/test_uint256.py]: Change this to 33. Always test the edge cases :)

File: tests/parser/types/numbers/test_uint256.py [hunk 3]
```diff
@@ -1,4 +1,43 @@
...
+from vyper.exceptions import (
+    InvalidLiteralException
+)
+
+
+def test_convert_bytes_to_uint256(assert_compile_failed, get_contract_with_gas_estimation):
+    # Test valid bytes input for conversion
+    test_success = """
+@public
+def foo(bar: bytes[5]) -> uint256:
```
Comment (fubuloubu) [inline tests/parser/types/numbers/test_uint256.py]: Can you add a test case that shows `bytes[32]`?

File: tests/parser/types/numbers/test_uint256.py (hunk 2 above)
Comment (jakerockland) [reply tests/parser/types/numbers/test_uint256.py]: Ah yes sorry will do, also will do on #1085

Comment (jakerockland) [reply tests/parser/types/numbers/test_uint256.py]: 😅

Comment (jakerockland) [reply tests/parser/types/numbers/test_uint256.py]: @fubuloubu Added better tests!

Comment (fubuloubu) [reply tests/parser/types/numbers/test_uint256.py]: Wait, is #1085 mixed commits with this one?

Comment (fubuloubu) [reply tests/parser/types/numbers/test_uint256.py]: Ah yes, I missed the fact that here you are testing uint256, and in the other one you are testing int128

**Comment by jakerockland** [comment]
Ah yeah haha they are easy to confuse 😅

**Comment by jakerockland** [comment]
@fubuloubu I believe I addressed the changes you requested and just made a commit to fix a merge conflict that came up.
